tasks/src/lib.rs

```diff
@@ -80,22 +135,114 @@ pub trait Task: 'static {
 }
 
 /// Scheduler managing enqueued tasks and routing incoming responses to active tasks.
+#[derive(Debug)]
 pub struct Scheduler {
     flow: ClientFlow,
     waiting_tasks: TaskMap,
     active_tasks: TaskMap,
     tag_generator: TagGenerator,
+    capabilities: Vec1<Capability<'static>>,
+    server_id: Option<Vec<(IString<'static>, NString<'static>)>>,
 }
 
 impl Scheduler {
-    /// Create a new scheduler.
-    pub fn new(flow: ClientFlow) -> Self {
-        Self {
+    pub async fn try_new_flow(
+        stream: AnyStream,
+        options: ClientFlowOptions,
+    ) -> Result<Self, SchedulerError> {
+        let (flow, greeting) = ClientFlow::receive_greeting(stream, options).await?;
+
+        let mut scheduler = Self {
             flow,
             waiting_tasks: Default::default(),
             active_tasks: Default::default(),
             tag_generator: TagGenerator::new(),
+            capabilities: Vec1::from(Capability::Imap4Rev1),
+            server_id: Default::default(),
+        };
+
+        // Capabilities may not be returned by the greeting. In this
+        // particular case, we explicitly request them.
+        scheduler.capabilities = if let Some(Code::Capability(capabilities)) = greeting.code {
+            capabilities
+        } else {
+            scheduler.capability().await?
+        };
+
+        Ok(scheduler)
+    }
+
+    pub async fn authenticate<'a>(
```

Comment:
This function also prevents users to have wrong or missing capabilities. It allows users to send an authenticate task, then get the right capabilities (either from the auth task itself, or from explicit request). It also takes care of sending client ids and storing server's ones.